microsoft/testsuites/cpu/common.py

```diff
@@ -0,0 +1,177 @@
+            target_cpu = channel_vp_map.target_cpu
+            if target_cpu == "0":
+                continue
+            cpu_in_used.add(target_cpu)
+
+    # get all cpu exclude cpu 0, usually cpu 0 is not allowed to do hotplug
+    cpu_count = node.tools[Lscpu].get_core_count()
+    all_cpu = list(range(1, cpu_count))
+
+    # get the idle cpu by excluding in used cpu from all cpu
+    idle_cpu = [str(x) for x in all_cpu if str(x) not in cpu_in_used]
+    return idle_cpu
+
+
+def set_idle_cpu_offline_online(log: Logger, node: Node, idle_cpu: List[str]) -> None:
+    for target_cpu in idle_cpu:
+        set_offline = set_cpu_state(node, target_cpu, False)
+        log.debug(f"set cpu{target_cpu} from online to offline.")
+        exception_message = (
+            f"expected cpu{target_cpu} state: {CPUState.OFFLINE}(offline), "
+            f"actual state: {CPUState.ONLINE}(online)."
+        )
+        if not set_offline:
+            raise BadEnvironmentStateException(
+                exception_message,
+                f"the test failed leaving cpu{target_cpu} in a bad state.",
+            )
+
+        set_online = set_cpu_state(node, target_cpu, True)
+        log.debug(f"set cpu{target_cpu} from offline to online.")
+        exception_message = (
+            f"expected cpu{target_cpu} state: {CPUState.ONLINE}(online), "
+            f"actual state: {CPUState.OFFLINE}(offline)."
+        )
+        if not set_online:
+            raise BadEnvironmentStateException(
+                exception_message,
+                f"the test failed leaving cpu{target_cpu} in a bad state.",
+            )
+
+
+def verify_cpu_hot_plug(log: Logger, node: Node, run_times: int = 1) -> None:
+    check_runnable(node)
+    file_path_list: Dict[str, str] = {}
+    restore_state = False
+    try:
+        for iteration in range(1, run_times + 1):
+            log.debug(f"start the {iteration} time(s) testing.")
+            restore_state = False
+            # set vmbus channels target cpu into 0 if kernel supports this feature.
+            file_path_list = set_vmbus_channels_target_cpu(log, node)
+            # when kernel doesn't support above feature, we have to rely on current vm's
+            # cpu usage. then collect the cpu not in used exclude cpu0.
+            idle_cpu = get_idle_cpu_set(node)
+            if 0 == len(idle_cpu):
+                raise SkippedException(
+                    "all of the cpu are associated vmbus channels,"
+                    " no idle cpu can be used to test hotplug."
+                )
+            # start to take idle cpu from online to offline, then offline to online.
+            set_idle_cpu_offline_online(log, node, idle_cpu)
+            # when kernel doesn't support set vmbus channels target cpu feature, the
+            # dict which stores original status is empty, nothing need to be restored.
+            restore_interrupts_assignment(file_path_list, node)
```

Comment:
It doesn't need it, the "finally" clause always executes.

Reply:
when test is under stress, this step is needed in every iteration, so that the next iteration is like a fresh run.